Flesh out the initial tracing SHIM API.

vext01 opened the pull request:
This is a DNMY PR. I'm raising for comments on the API design of yktrace.

Please see the commit message for more info.

If we like the design, there's a change to ykrustc that needs to go in first, then this PR, then we should enable buildbot on this repo.

Cheers

vext01 committed:
Flesh out the initial tracing SHIM API.

Although for now ykrustc only supports software tracing, I've designed
the API assuming that there can be a runtime switch between different
tracing backends (e.g. hardware/software tracing). Making ykrustc
support such a switch is future work.

For now the API supports:
 * Starting a trace and returning a thread tracer struct.
 * Stopping a trace, consuming the thread tracer (to prevent double-stop
   scenarios), and returning the recorded trace.
 * Getting a MIR location from a trace at a given index.
 * Iteration over a trace.

The API also adds a Drop implementation for software traces so that
users don't have to manually free a raw pointer at the correct time.

vext01 commented:
This is everything apart from the buildbot.

NOTE: we will have to revise the commit message before merge (handles -> tracers).

ltratt commented on src/lib.rs:
```diff
@@ -0,0 +1,191 @@
+pub fn start_tracing(kind: TracingKind) -> Box<dyn ThreadTracer> {
+    match kind {
+        TracingKind::SoftwareTracing => swt::start_tracing(),
+        _ => panic!("tracing kind not implemented")
```
Should be `unimplemented`?

vext01 replied:
yeah, probably.

ltratt commented on src/swt.rs:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    fn loc(&self, idx: usize) -> &MirLoc {
+        if idx > isize::max_value() as usize {
```
Why are we checking against `isize::max_value`? Is it because we're using `offset` below? If so, `add` negates the need for this check (assuming, as I think we must, that `ptr + self.len` is in bounds).

vext01 replied:
> Is it because we're using offset below?

Yeah, offset accepts an isize so that you can bump pointers both back and forward:
```
pub unsafe fn offset(self, count: isize) -> *const T
```

> pub unsafe fn offset(self, count: isize) -> *const T

Ah! I did not know about `Ptr::add`! Marvellous. Will fix.

ltratt commented on src/swt.rs:
```diff
@@ -0,0 +1,106 @@
+
+impl ThreadTracer for SWTThreadTracer {
+    fn stop_tracing(self: Box<Self>) -> Option<Box<dyn MirTrace>> {
+        match yk_swt::stop_tracing() {
```
Maybe `yk_swt::stop_tracing().map(|(buf, len)| Some(Box::new(...))`?

vext01 replied:
Ah yes. I never was good at all of these combinators :)

ltratt commented on src/lib.rs:
```diff
@@ -0,0 +1,191 @@
+/// A thread tracer is returned through which tracing can be stopped and the trace obtained. The
+/// user must ensure that only one thread tracer exists at any one time per-thread, otherwise
+/// undefined behaviour will result.
+pub fn start_tracing(kind: TracingKind) -> Box<dyn ThreadTracer> {
```
Is there a function somewhere which says "start tracing using whatever the best available tracer is"?

vext01 replied:
I've not implemented one, no. We could have `kind` an `Option` where `None` means "just frigging choose one, will ya!".

ltratt replied:
I was thinking the same thing. Let's do it!

vext01 commented:
The `rustfmt` hack I used yesterday no longer works. Trying to figure out how to do it properly...

vext01 commented:
> The rustfmt hack I used yesterday no longer works. Trying to figure out how to do it properly...

For future reference, to build rustfmt from `x.py` it's: `./x.py build src/tools/rustfmt` then you have to run the tool manually, i.e.: `~/research/ykrustc_pv/build/x86_64-unknown-linux-gnu/stage2-tools-bin/rustfmt  src/swt.rs`.

You can't build rustfmt at stage1, and once it's built in this way you can't run it via `cargo fmt` either.

Ready for re-review.

vext01 commented:
Ready.

ltratt commented:
Please squash.

vext01 commented:
I forgot to update the docstring. If it looks good, I'll squash.

ltratt commented on src/lib.rs:
```diff
@@ -74,9 +74,10 @@ pub trait ThreadTracer {
-/// A thread tracer is returned through which tracing can be stopped and the trace obtained. The
-/// user must ensure that only one thread tracer exists at any one time per-thread, otherwise
-/// undefined behaviour will result.
+/// By passing `kind` as `None` an appropriate tracing kind will be selected automatically. A
```
How about simplifying to something like "If `None` is passed, then an appropriate backend will be selected; if `Some(...)` is passed, a specific backend can be specified. Only one thread tracer can be started at any point; starting more than one leads to undefined behaviour."?

vext01 replied:
Try my last push. I want the restrictions on threads to be super-clear.

ltratt commented on src/lib.rs:
```diff
@@ -74,10 +74,9 @@ pub trait ThreadTracer {
-/// must ensure that only one thread tracer exists at any one time per-thread, otherwise undefined
-/// behaviour will result.
+/// If `None` is passed, then an appropriate tracing kind will be selected; by passing `Some(...)`,
+/// a specific kind can be chosen. Tracing different threads concurrently is allowed, but tracing
```
The second sentence doesn't really help IMHO. Maybe something like "Any given thread can only have one active tracer; calling `start_tracing` on a thread where there is already an active tracer leads to undefined behaviour".

vext01 replied:
I like it!

ltratt replied:
"can only have one active tracer" might better be "must have at most one active tracer"?

vext01 replied:
I wrote exactly the same independently!

vext01 commented:
Try this.

ltratt commented:
Please squash.

vext01 commented:
Squashed and revised commit message.

Before we merge this, we need to review and merge: https://github.com/softdevteam/ykrustc/pull/5

vext01 commented:
With the ykrustc PR now merged we can now do this one.

My next PR will enable buildbot and bors on this repo.

ltratt commented:
Why don't we do the buildbot/bors PR first, merge that into master; then we can merge master into this PR and use buildbot to check this PR?

ltratt commented:
Why don't we do the buildbot/bors PR first, merge that into master; then we can merge master into this PR and use buildbot to check this PR?

vext01 commented:
> Why don't we do the buildbot/bors PR first

We could, I'd have to add some dummy code to build though.

ltratt commented:
Maybe we should just add the buildbot stuff into this PR as a separate commit? It's fairly self-contained, right?

vext01 commented:
> Maybe we should just add the buildbot stuff into this PR as a separate commit? It's fairly self-contained, right?

We can try, but like i said the other day, I'm not sure if bors will see existing PRs. It's something I'd like to know, so no objections to trying.

ltratt commented:
Let's try it then :)

vext01 commented:
OK, it may take a little while, as I'm struggling to get a working rustfmt and rustdoc in our ykrustc image.

vext01 committed:
Enable buildbot/bors.

vext01 commented:
bors ping

vext01 commented:
bors try

ltratt commented:
It doesn't look like the build has started?

vext01 commented:
I'm working on it.

vext01 commented:
bors try-

vext01 commented:
bors try

vext01 commented:
bors try

vext01 commented:
OK, hopefully this should work.

I made the mistake of informing the buildbot about the new repo after `bors try`, which meant buildbot saw the `trying` commit as "old news" and didn't bother trying to build it.

The build will take ages because we have to build a full stage2 compiler to have rustdoc and rustfmt (the build targets for these fail if you try to force stage 1 only).

In the meantime, are you happy with the CI commits?

ltratt commented:
I sort-of assumed you'd want to squash the third commit (the merge) away? If so, do you want to squash now, then we can do a full build (an r plus build).

vext01 commented:
> I sort-of assumed you'd want to squash the third commit 

Well yes, but you have to request squashes.

Since you are OK with it, I'll do it now.

vext01 commented:
Squashed

ltratt commented:
bors r+

vext01 commented:
> run pkg_config fail: "Failed to run `\"pkg-config\" \"--libs\" \"--cflags\" ...

This is due to missing packages on the buildbot machine. I've added them (for the record `libssl-dev` and `pkg-config`).

I've also added the `config.toml` options from ykrustc. Most importantly, use threads to build -- should be faster.

Shall I squash it in?

ltratt commented:
Yes, please squash.

vext01 commented:
splat

ltratt commented:
bors r+

ltratt commented:
Hmm, 9 hours is a long time. Let's try again.

bors r+
